src/auth/stscredentials.ts

```diff
@@ -0,0 +1,179 @@
+   * @param clientAuthentication The client authentication credentials if
+   *   available.
+   */
+  constructor(
+    private readonly tokenExchangeEndpoint: string,
+    clientAuthentication?: ClientAuthentication
+  ) {
+    super(clientAuthentication);
+    this.transporter = new DefaultTransporter();
+  }
+
+  /**
+   * Exchanges the provided token for another type of token based on the
+   * rfc8693 spec.
+   * @param stsCredentialsOptions The token exchange options used to populate
+   *   the token exchange request.
+   * @param additionalHeaders Optional additional headers to pass along the
+   *   request.
+   * @param options Optional additional GCP-specific non-spec defined options
+   *   to send with the request.
+   *   Example: `&options=${encodeUriComponent(JSON.stringified(options))}`
+   * @return A promise that resolves with the token exchange response containing
+   *   the requested token and its expiration time.
+   */
+  async exchangeToken(
+    stsCredentialsOptions: StsCredentialsOptions,
+    additionalHeaders?: Headers,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    options?: {[key: string]: any}
+  ): Promise<StsSuccessfulResponse> {
+    const values: StsRequestOptions = {
+      grant_type: stsCredentialsOptions.grantType,
+      resource: stsCredentialsOptions.resource,
+      audience: stsCredentialsOptions.audience,
+      scope: stsCredentialsOptions.scope?.join(' '),
+      requested_token_type: stsCredentialsOptions.requestedTokenType,
+      subject_token: stsCredentialsOptions.subjectToken,
+      subject_token_type: stsCredentialsOptions.subjectTokenType,
+      actor_token: stsCredentialsOptions.actingParty?.actorToken,
+      actor_token_type: stsCredentialsOptions.actingParty?.actorTokenType,
+      // Non-standard GCP-specific options.
+      options: options && JSON.stringify(options),
+    };
+    // Remove undefined fields.
+    Object.keys(values).forEach(key => {
+      if (typeof values[key] === 'undefined') {
+        delete values[key];
+      }
+    });
+
+    const headers = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    };
+    // Inject additional STS headers if available.
+    Object.assign(headers, additionalHeaders || {});
+
+    const opts: GaxiosOptions = {
+      url: this.tokenExchangeEndpoint,
+      method: 'POST',
+      headers,
+      data: querystring.stringify(values),
+      responseType: 'json',
+    };
+    // Apply OAuth client authentication.
+    this.applyClientAuthenticationOptions(opts);
+
+    try {
+      const response = await this.transporter.request<StsSuccessfulResponse>(
+        opts
+      );
+      // Successful response.
+      const stsSuccessfulResponse = response.data;
+      stsSuccessfulResponse.res = response;
+      return stsSuccessfulResponse;
+    } catch (error) {
+      // Translate error to OAuthError.
+      if (error.response) {
+        throw getErrorFromOAuthErrorResponse(
```

Comment:
Just to double check - does this method create a `new Error`, or modify the message on the existing?  I want to be about not creating a new error, because we want to preserve the original stack trace.

Reply:
It was creating a new error and not preserving the original error data. I have extended this to preserve the original error data (including stack) but to modify the error message.